### PR TITLE
Compass heading jitter fix

### PR DESCRIPTION
**Pull request by tcobbs-bentley**

### CODE REVIEW

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMGeolocationManager.kt [hunk 1]
```diff
@@ -358,17 +393,54 @@ class ITMGeolocationManager(private var context: Context) {
...
+                }
+            }
+        }
+        if (headingSensor == null && rotationSensor == null) {
+            // Both types of rotation sensor produce the same data, so use the same variables for
+            // both.
+            rotationSensor = sensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR)
+            if (rotationSensor == null) {
+                rotationSensor = sensorManager.getDefaultSensor(Sensor.TYPE_GEOMAGNETIC_ROTATION_VECTOR)
+            }
```
Comment (toddsouthenbentley) [inline mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMGeolocationManager.kt:424]:
```suggestion
            rotationSensor = sensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR) ?:
                sensorManager.getDefaultSensor(Sensor.TYPE_GEOMAGNETIC_ROTATION_VECTOR)
```

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMGeolocationManager.kt [hunk 2]
```diff
@@ -358,17 +393,54 @@ class ITMGeolocationManager(private var context: Context) {
...
+        // frequently than that. So a timer gets used in the callbacks to prevent a deluge of
+        // updates from being sent to JS. The timer sends updates a maximum of 10 times per second.
+        sensorManager.registerListener(sensorListener, accelerometerSensor, SensorManager.SENSOR_DELAY_UI)
+        listening = true
+        if (headingSensor != null) {
+            sensorManager.registerListener(sensorListener, headingSensor, SensorManager.SENSOR_DELAY_UI)
+        } else if (rotationSensor != null) {
+            sensorManager.registerListener(sensorListener, rotationSensor, SensorManager.SENSOR_DELAY_UI)
+        } else if (magneticSensor != null) {
+            sensorManager.registerListener(sensorListener, magneticSensor, SensorManager.SENSOR_DELAY_UI)
```
Comment (toddsouthenbentley) [inline mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMGeolocationManager.kt:443]:
```suggestion
        sensorManager.registerListener(sensorListener, headingSensor ?: rotationSensor ?: magneticSensor, SensorManager.SENSOR_DELAY_UI)
```

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMGeolocationManager.kt [hunk 3]
```diff
@@ -381,53 +453,142 @@ class ITMGeolocationManager(private var context: Context) {
...
+                magneticHeadings.forEach {
+                    coord += Coord2D.fromDegrees(it)
+                }
+                // Divide by magnitude to get a unit vector pointing in the right direction
+                coord /= coord.magnitude()
+                // Make sure our average coordinate is on the unit circle
+                assert(abs(coord.magnitude() - 1.0) < 0.0001)
+                // Convert back to polar coordinates
+                heading = (Math.toDegrees(atan2(coord.y, coord.x)) + 360.0) % 360.0
+                magneticHeadings.clear()
```
Comment (toddsouthenbentley) [inline mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMGeolocationManager.kt:590]: For readability, I suggest abstracting this code to a new private function.

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMGeolocationManager.kt [hunk 4]
```diff
@@ -381,53 +453,142 @@ class ITMGeolocationManager(private var context: Context) {
...
             return null
         }
-        val rotationMatrixB = FloatArray(9)
-        val (axisX, axisY) = getHeadingAxes() ?: return null
-        if (!SensorManager.remapCoordinateSystem(rotationMatrixA, axisX, axisY, rotationMatrixB)) {
+        return getHeadingFromRotationMatrix(rotationMatrix, isUpsideDown)
+    }
+
+    private fun clampHeading(heading: Double?): Double? = if (heading != null) {
+        val clamp = heading % 360.0
```
Comment (toddsouthenbentley) [inline mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMGeolocationManager.kt:524]: Can this just simply be: `clamp = (heading + 360.0) % 360.0`?

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMGeolocationManager.kt (hunk 2 above)
Comment (toddsouthenbentley) [reply mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMGeolocationManager.kt:443]: Actually, now that I think about it, do we really need the 3 sensor variables (magnetic, rotation and heading)? Or can we just have one and set it according to our preferred order?

Comment (tcobbs-bentley) [reply mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMGeolocationManager.kt]: I refactored the code to only use two sensor variables (accelerometer + other).

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMGeolocationManager.kt (hunk 3 above)
Comment (tcobbs-bentley) [reply mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMGeolocationManager.kt]: Fixed.

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMGeolocationManager.kt (hunk 4 above)
Comment (tcobbs-bentley) [reply mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMGeolocationManager.kt:511]: I had something like that originally, but realized that if something caused the value to get highly negative (< -360.0), it wouldn't work. I don't _think_ the value going in can be less than -360.0, but I'm not positive about that. Also, the function is named "clampHeading", so it should work with any possible input value.

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMGeolocationManager.kt [hunk 5]
```diff
@@ -354,22 +387,48 @@ class ITMGeolocationManager(private var context: Context) {
...
     }
 
+    private fun getHeadingSensor(): Sensor? {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+            // We don't have any test phones that return non-null for Sensor.TYPE_HEADING.
+            // Until we do, we have no idea if our processing of the data is correct. So,
+            // while the code to interpret the data from Sensor.TYPE_HEADING is being left
+            // in the library, the following line makes sure that that data is never used.
+            // Add return before the line below to activate usage of the heading sensor.
+            sensorManager.getDefaultSensor(Sensor.TYPE_HEADING)
```
Comment (toddsouthenbentley) [inline mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMGeolocationManager.kt:397]: The comment here indicates we should be returning null and the `getDefaultSensor` call should be commented out for now.

Comment (tcobbs-bentley) [reply mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMGeolocationManager.kt]: Fixed (hopefully)

**Review by toddsouthenbentley — APPROVED** [review]

**Pull request merged**